axelrod/strategies/finite_state_machines.py

```diff
@@ -1,9 +1,347 @@
+    2:D;C | Back-trace: _ | On-state: 2
+    3:C;C | Back-trace: _ | On-state: 3
+    3:D;D | Back-trace: _ | On-state: 3
+
+    In the second step, we walk backwards along incoming transitions.
+    We continue to label each branch by its ending
+    state:previous_opponent_action;next_action, but we list the state that
+    we're on at this point in time:
+    1:C;D | Back-trace: _/D, C/_ | On-state: 1
+    1:C;D | Back-trace: _/D, C/_ | On-state: 2
+    1:C;D | Back-trace: _/D, C/_ | On-state: 3
+    2:C;D | Back-trace: _/D, C/_ | On-state: 1
+    1:D;D | Back-trace: _/D, D/_ | On-state: 1
+    1:D;D | Back-trace: _/D, D/_ | On-state: 2
+    1:D;D | Back-trace: _/D, D/_ | On-state: 3
+    2:D;C | Back-trace: _/D, D/_ | On-state: 1
+    3:C;C | Back-trace: _/C, C/_ | On-state: 2
+    3:C;C | Back-trace: _/C, C/_ | On-state: 3
+    3:D;D | Back-trace: _/C, D/_ | On-state: 2
+    3:D;D | Back-trace: _/C, D/_ | On-state: 3
+
+    From here we can conclude that:
+    If _/D, C/_, then D
+    If _/C, C/_, then C
+    If _/C, D/_, then D
+
+    We remove the branches that correspond to those action chains.  But we
+    continue to walk back the _/D, D/_ branches:
+    1:D;D | Back-trace: _/D, C/D, D/_ | On-state: 1
+    1:D;D | Back-trace: _/D, C/D, D/_ | On-state: 2
+    1:D;D | Back-trace: _/D, C/D, D/_ | On-state: 3
+    1:D;D | Back-trace: _/D, C/D, D/_ | On-state: 1
+    1:D;D | Back-trace: _/C, D/D, D/_ | On-state: 2
+    1:D;D | Back-trace: _/C, D/D, D/_ | On-state: 3
+    2:D;C | Back-trace: _/D, D/D, D/_ | On-state: 1
+    2:D;C | Back-trace: _/D, D/D, D/_ | On-state: 2
+    2:D;C | Back-trace: _/D, D/D, D/_ | On-state: 3
+
+    From here we conclude that:
+    If _/D, C/D, D/_, then D
+    If _/C, D/D, D/_, then D
+    If _/D, D/D, D/_, then C
+
+    There are no more undecided branches, so we stop and say that the memory
+    is 2.
+    """
+    # If initial_state is set, use this to determine which transitions are
+    # reachable from the initial_state and restrict to those.
+    if initial_state is not None:
+        transitions = get_accessible_transitions(transitions, initial_state)
+
+    # First make a back_transitions dict from transitions.  This is keyed on
+    # states, and a list of "BackTrans" (one for each transition incoming to
+    # that state) as values.
+    back_transitions = defaultdict(list)
+    # A "BackTrans" has the previous state and previous action/reaction pair.
+    BackTrans = namedtuple("BackTrans", ["prev_state", "prev_reaction",
+                                         "prev_opp_action"])
+    for trans in transitions:
+        state = trans[0]
+        last_opponent_action = trans[1]
+        next_state = trans[2]
+        next_action = trans[3]
+
+        back_transitions[next_state].append(BackTrans(state,
+                                                      next_action,
+                                                      last_opponent_action))
+
+    class ActionChain(object):
```

Comment:
Maybe move these classes out of the function, it doesn't seem like they use any local data in their definitions.

Reply:
You're right that they could be moved outside of the function without problem.  My reasoning for leaving them in the function is because they're not used anywhere else, and my feeling is that it's similar to a local variable, which could be made global, but encapsulation is better.  I know this is a contentious point though.

Reply:
I'd prefer to see these classes outside the function and with unit tests for their behaviour.